File: 2/KadanesAlgorithm.java
class KadanesAlgorithm {
	public static int kadanesAlgorithm(int arr[]) {
		int max = arr[0];
		int maxTillHere = arr[0];	

		for(int i=1; i<arr.length; i++) {
			maxTillHere = Math.max(maxTillHere+arr[i], arr[i]);
			max = Math.max(max, maxTillHere);
		}
		
		return max;
	}

	public static void main(String args[]) {
		int arr[] = {3,5,-9,1,3,-2,3,4,7,2,-9,6,3,1,-5,4};
		System.out.println(kadanesAlgorithm(arr));
	}
}

File: 2/js/src/KadanesAlgorithm.ts
function solve(list: number[]): number {
	const n = list.length;
	let max = list[0];

	for(let i=1; i<n; i++) {
		list[i] = Math.max(list[i], list[i] + list[i-1]);
		max = Math.max(list[i], max);
	}

	return max;
}

(function main() {
	const list = [3,5,-9,1,3,-2,3,4,7,2,-9,6,3,1,-5,4];

	console.log(solve(list));
})();
